src/MongoDB.Driver/Encryption/AutoEncryptionLibMongoController.cs

```
@@ -153,7 +203,7 @@ protected override async Task ProcessStateAsync(CryptContext context, string dat
         // private methods
         private void ProcessNeedCollectionInfoState(CryptContext context, string databaseName, CancellationToken cancellationToken)
         {
-            var database = _client.GetDatabase(databaseName);
+            var database = _metadataClient.GetDatabase(databaseName);
```

I will add isNotNull validation here in the next commit

---

It's usually not necessary to add this kind of validation in private methods.

If we are the only ones calling this method we can make sure that it is not called when `_metadataClient` is null.

You can leave this if you want... but let's not be tempted to start adding argument validation to all our private methods.

For public methods of course we have to validate arguments.

---

we don't have control over when this method will be called since it's up to the libmongocrypt level logic. And we definitely can have null in the metadataClient. So, it's about preventing a null reference exception in case if something will be changed on the C level. So, I would say in this particular case it's required.

---

but I agree that this should not be a common rule

---

We do know whether this method will be called or not. If `AutoEncryptionOptions.BypassAutoEncryption` is `true` then we know that this method is never called.

That's precisely why we set `_metadataClient` to `null` when we know it won't be needed.

---

>We do know whether this method will be called or not. If AutoEncryptionOptions.BypassAutoEncryption is true then we know that this method is never called.

if there is no bug in the C level, then yes, but we don't have control over this. It can be changed at any moment, we can have a bug in this logic even now, and we may not see this just because it happens only in some circumstances

---

I don't say that this is expected, but I do say that libmongocrypt is an external library for the driver and it's better to validate their "output"

---

`_metadataClient` is not output of the libmongocrypt library.

We create it ourselves (unless `AutoEncryptionOptions.BypassAutoEncryption` is `true`).

As I understand it the C level can't call us back unless we call it first (which we won't do if `BypassAutoEncryption` is `true`).

Anyway... if you feel better with this `Ensure` here I'm still OK with you leaving it.

---

Note: `Ensure` is used for validating arguments... and therefore throws an `ArgumentNullException` here. But `_metadataClient` is not an argument, it's part of the object state.

Perhaps you should manually check for null and throw an `InvalidOperationException` instead.

---

the C level can potentially call this method even in decrypt steps, so in that case, this value will be null

---

>Perhaps you should manually check for null and throw an InvalidOperationException instead.

yeah, I also thought about this, I will do it

---

> the C level can potentially call this method even in decrypt steps, so in that case, this value will be null

Are you sure?

If that's true then why are we setting `_metadataClient` to `null` when `BypassAutoEncryption` is `true`?

---

I meant that it may be if there is a bug on the C level (with "routing"). But it's not expected according to the spec

---

>>    Perhaps you should manually check for null and throw an InvalidOperationException instead.

>yeah, I also thought about this, I will do it

done